## Prediction Model based on Genres


### Model 1: generate list of movie recommendation based on "genres" + "rating".
- Input from user: genres
- scores are calculated and ranked to produce the top 5 movies
- In this notebook, I will offer 2 simple way:
    - based on rating score alone
    - based on rating score and genre picked by the user
- Features include: 
    - genres
    - movie rating score (based on IMDB formula)

In [227]:
import pandas as pd

pd.set_option("display.max_columns", 85)
pd.set_option("display.max_rows", 85)
df = pd.read_csv('../../edit_data/Lee/cleaned_data/movies_main.csv')  
df.head()

Id               Original_Title Original_Language  \
0    862                    Toy Story                en   
1   8844                      Jumanji                en   
2  15602             Grumpier Old Men                en   
3  31357            Waiting to Exhale                en   
4  11862  Father of the Bride Part II                en   

                                    Spoken_Languages      Budget      Revenue  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  30000000.0  373554033.0   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  65000000.0  262797249.0   
2           [{'iso_639_1': 'en', 'name': 'English'}]         0.0          0.0   
3           [{'iso_639_1': 'en', 'name': 'English'}]  16000000.0   81452156.0   
4           [{'iso_639_1': 'en', 'name': 'English'}]         0.0   76578911.0   

   Runtime Release_Date                               Production_Companies  \
0     81.0   1995-10-30     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1    104.0   1995-12-15  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2    101.0   1995-12-22  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3    127.0   1995-12-22  [{'name': 'Twentieth Century Fox Film Corporat...   
4    106.0   1995-02-10  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                Production_Countries    imdb_id  Popularity  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114709   21.946943   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113497   17.015539   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113228   11.712900   
3  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114885    3.859495   
4  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113041    8.387519   

   Vote_Average  Vote_Count  \
0           7.7      5415.0   
1           6.9      2413.0   
2           6.5        92.0   
3           6.1        34.0   
4           5.7       173.0   

                               Belongs_To_Collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...   
3                                                NaN   
4  {'id': 96871, 'name': 'Father of the Bride Col...   

                                             Tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                                            Overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                              Genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                         Genres_Parse         Belongs_To_Collection_Parse  \
0   ['Animation', 'Comedy', 'Family']            ['Toy Story Collection']   
1  ['Adventure', 'Fantasy', 'Family']                                 NaN   
2               ['Romance', 'Comedy']       ['Grumpy Old Men Collection']   
3      ['Comedy', 'Drama', 'Romance']                                 NaN   
4                          ['Comedy']  ['Father of the Bride Collection']   

    Spoken_Languages_Parse                         Production_Companies_Parse  \
0              ['English']                        ['Pixar Animation Studios']   
1  ['English', 'Français']  ['TriStar Pictures', 'Teitler Fi

Method 1: step 1 - calculate the C and m values.

In [231]:
## To calculate the mean value of Vote Average
C= df['Vote_Average'].mean()
C

5.6259312895463855

In [232]:
## set vote count threshold to 90 %tile.
m= df['Vote_Count'].quantile(0.9)
m

163.0

Method 1: Step 2 - Set qualified Movies (The Filter)
- Qualified movies are set based on Number of Vote Count received by the movies. 
- To qualified, the movies must has be at 90 %tile in total vote counts (i.e. 163 votes or more)
- This is where we would set the ['Genres']' as a filter. 

Filter 1: m-value only

In [110]:
## only filter based on m value
# qualified_movies = df.copy().loc[df['Vote_Count'] >= m]
# qualified_movies.shape


Filter 2: m-value + genres (entered by user)

Re-Parse Genres again to a new column (the parsed column is not working)

In [111]:
import ast
import numpy as np

# Function to parse the string and extract 'name'
def extract_names(data_string):
    if pd.isna(data_string):
        return np.nan
    try:
        # Safely evaluate the string to a Python object
        data_object = ast.literal_eval(data_string)
        if isinstance(data_object, list):
            # Extract 'name' from each dictionary in the list
            names = [item['name'] for item in data_object]
            return names
        elif isinstance(data_object, dict):
            # Extract 'name' from the dictionary
            return [data_object.get('name', np.nan)]
    except (ValueError, SyntaxError):
        return np.nan
    
    
df['Genres_Parse_new'] = df['Genres'].apply(extract_names)

In [112]:
### TEST : to generate a unique list of genres

# Flatten the list of genres, ignoring NaN values, and extract unique values
unique_genres = set(
    genre
    for sublist in df['Genres_Parse_new'].dropna()  # Drop NaN values
    for genre in sublist
)

# Convert the set to a sorted list (optional)
unique_genres_list = sorted(unique_genres)

# Display the unique genres
print(unique_genres_list)
print(len(unique_genres_list))

['Action', 'Adventure', 'Animation', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Family', 'Fantasy', 'Foreign', 'History', 'Horror', 'Music', 'Mystery', 'Romance', 'Science Fiction', 'TV Movie', 'Thriller', 'War', 'Western']
20


Request user to enter the Genre
- codes can be enhanced to avoid typo, etc. 

In [113]:

# Genres List in total (20)
genres_list= ['Action', 'Adventure', 'Animation', 
              'Comedy', 'Crime', 'Documentary', 
              'Drama', 'Family', 'Fantasy', 
              'Foreign', 'History', 'Horror', 
              'Music', 'Mystery', 'Romance', 
              'Science Fiction', 'TV Movie', 
              'Thriller', 'War', 'Western']

print('total genres:', len(genres_list))

genre_input = input("Enter the genre: ")

if x in genres_list:
    filt = (df['Vote_Count'] >= m) & (df['Genres_Parse_new'].apply(lambda x: genre_input in x if isinstance(x, list) else False) )
    qualified_movies = df.copy().loc[filt]
    print("You have selected the genre: ", genre_input)
    print("qualified_movies: ", qualified_movies.shape)
else:
    print("wrong genre entered.")
    Game_is_on = False



total genres: 20
You have selected the genre:  Science Fiction
qualified_movies:  (578, 29)


Method 1: step 3 - set up the rating score calculation

In [230]:
def weighted_rating(x, m=m, C=C):
    v = x['Vote_Count']
    R = x['Vote_Average']
    # Calculation based on the IMDB formula
    return (v/(v+m) * R) + (m/(m+v) * C)

In [115]:
# Define a new feature 'score' and calculate its value with `weighted_rating()`
qualified_movies['Score'] = qualified_movies.apply(weighted_rating, axis=1)

qualified_movies.head()

Id              Original_Title Original_Language  \
28     902  La Cité des Enfants Perdus                fr   
31      63              Twelve Monkeys                en   
157  10329                       Congo                en   
169   9886             Johnny Mnemonic                en   
170   9482                 Judge Dredd                en   

                                      Spoken_Languages      Budget  \
28   [{'iso_639_1': 'cn', 'name': '广州话 / 廣州話'}, {'i...  18000000.0   
31   [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  29500000.0   
157  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  50000000.0   
169  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  25000000.0   
170           [{'iso_639_1': 'en', 'name': 'English'}]  90000000.0   

         Revenue  Runtime Release_Date  \
28     1738611.0    108.0   1995-05-16   
31   168840000.0    129.0   1995-12-29   
157  152022101.0    109.0   1995-06-09   
169   19075720.0     97.0   1995-05-26   
170  113493481.0     96.0   1995-06-30   

                                  Production_Companies  \
28   [{'name': 'Procirep', 'id': 311}, {'name': 'Co...   
31   [{'name': 'Universal Pictures', 'id': 33}, {'n...   
157  [{'name': 'Paramount Pictures', 'id': 4}, {'na...   
169  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
170  [{'name': 'Hollywood Pictures', 'id': 915}, {'...   

                                  Production_Countries    imdb_id  Popularity  \
28   [{'iso_3166_1': 'FR', 'name': 'France'}, {'iso...  tt0112682    9.822423   
31   [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114746   12.297305   
157  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0112715    7.260574   
169  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  tt0113481   11.715868   
170  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113492    8.184815   

     Vote_Average  Vote_Count Belongs_To_Collection  \
28            7.6       308.0                   NaN   
31            7.4      2470.0                   NaN   
157           5.0       214.0                   NaN   
169           5.5       380.0                   NaN   
170           5.4       643.0                   NaN   

                                               Tagline  \
28           Where happily ever after is just a dream.   
31                              The future is history.   
157              Where you are the endangered species.   
169  The hottest data on earth. In the coolest head...   
170                 In the future, one man is the law.   

                                              Overview  \
28   A scientist in a surrealist society kidnaps ch...   
31   In the year 2035, convict James Cole reluctant...   
157  Eight people embark on an expedition into the ...   
169  A data courier, literally carrying a data pack...   
170  In a dystopian future, Dredd, the most famous ...   

                                                Genres  \
28   [{'id': 14, 'name': 'Fantasy'}, {'id': 878, 'n...   
31   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
157  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
169  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
170           [{'id': 878, 'name': 'Science Fiction'}]   

                                          Genres_Parse  \
28         ['Fantasy', 'Science Fiction', 'Adventure']   
31          ['Science Fiction', 'Thriller', 'Mystery']   
157  ['Action', 'Adventure', 'Drama', 'Mystery', 'S...   
169  ['Adventure', 'Action', 'Drama', 'Science Fict...   
170                                ['Science Fiction']   

    Belongs_To_Collection_Parse     Spoken_Languages_Parse  \
28                          NaN  ['广州话 / 廣州話', 'Français']   
31                          NaN    ['English', 'Français']   
157                         NaN    ['English', 'Français']   
169                         NaN         ['English', '日本語']   
170                         NaN                ['English']   

                            Productio

Method 1: step 4 - create the movie recommendation based on the Score

In [193]:
#Sort movies based on score calculated above
qualified_movies = qualified_movies.sort_values('Score', ascending=False)

#Print the top 5 movies
print("Genre you have chosen:      ", genre_input)
print("Number of qualified movies: ", qualified_movies.shape)
qualified_movies[['Original_Title', 'Genres_Parse_new','Vote_Count','Vote_Average', 'Popularity','Score']].head(5)

Genre you have chosen:       Science Fiction
Number of qualified movies:  (578, 30)


Original_Title  \
1149   The Empire Strikes Back   
15395                Inception   
22702             Interstellar   
255                  Star Wars   
1220        Back to the Future   

                                        Genres_Parse_new  Vote_Count  \
1149                [Adventure, Action, Science Fiction]      5998.0   
15395  [Action, Thriller, Science Fiction, Mystery, A...     14075.0   
22702                [Adventure, Drama, Science Fiction]     11187.0   
255                 [Adventure, Action, Science Fiction]      6778.0   
1220        [Adventure, Comedy, Science Fiction, Family]      6239.0   

       Vote_Average  Popularity     Score  
1149            8.2   19.470959  8.131899  
15395           8.1   29.108149  8.071676  
22702           8.1   32.213481  8.064469  
255             8.1   42.149697  8.041900  
1220            8.0   25.778509  7.939554

### Model 2: Simple Linear KNN Model by Score

- KNN Regression Model
- input (features, X):
    - ['Vote_Average']
    - [''Vote_Count']
    - ['Budget']
    - ['Revenue']
    - [-'Runtime']
- Output (Target, y): ['Score']

In [228]:
# load the data
df.head()

Id               Original_Title Original_Language  \
0    862                    Toy Story                en   
1   8844                      Jumanji                en   
2  15602             Grumpier Old Men                en   
3  31357            Waiting to Exhale                en   
4  11862  Father of the Bride Part II                en   

                                    Spoken_Languages      Budget      Revenue  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  30000000.0  373554033.0   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  65000000.0  262797249.0   
2           [{'iso_639_1': 'en', 'name': 'English'}]         0.0          0.0   
3           [{'iso_639_1': 'en', 'name': 'English'}]  16000000.0   81452156.0   
4           [{'iso_639_1': 'en', 'name': 'English'}]         0.0   76578911.0   

   Runtime Release_Date                               Production_Companies  \
0     81.0   1995-10-30     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1    104.0   1995-12-15  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2    101.0   1995-12-22  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3    127.0   1995-12-22  [{'name': 'Twentieth Century Fox Film Corporat...   
4    106.0   1995-02-10  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                Production_Countries    imdb_id  Popularity  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114709   21.946943   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113497   17.015539   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113228   11.712900   
3  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114885    3.859495   
4  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113041    8.387519   

   Vote_Average  Vote_Count  \
0           7.7      5415.0   
1           6.9      2413.0   
2           6.5        92.0   
3           6.1        34.0   
4           5.7       173.0   

                               Belongs_To_Collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...   
3                                                NaN   
4  {'id': 96871, 'name': 'Father of the Bride Col...   

                                             Tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                                            Overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                              Genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                         Genres_Parse         Belongs_To_Collection_Parse  \
0   ['Animation', 'Comedy', 'Family']            ['Toy Story Collection']   
1  ['Adventure', 'Fantasy', 'Family']                                 NaN   
2               ['Romance', 'Comedy']       ['Grumpy Old Men Collection']   
3      ['Comedy', 'Drama', 'Romance']                                 NaN   
4                          ['Comedy']  ['Father of the Bride Collection']   

    Spoken_Languages_Parse                         Production_Companies_Parse  \
0              ['English']                        ['Pixar Animation Studios']   
1  ['English', 'Français']  ['TriStar Pictures', 'Teitler Fi

In [ ]:
### NOT USING THIS 
#create a mapping of unique title and Id
# lookup_title = dict(zip(df.Id.unique(), df.Original_Title.unique()))
# lookup_title

In [233]:
# add Score as a new column
df['Score'] = df.apply(weighted_rating, axis=1)
df.head()

Id               Original_Title Original_Language  \
0    862                    Toy Story                en   
1   8844                      Jumanji                en   
2  15602             Grumpier Old Men                en   
3  31357            Waiting to Exhale                en   
4  11862  Father of the Bride Part II                en   

                                    Spoken_Languages      Budget      Revenue  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  30000000.0  373554033.0   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  65000000.0  262797249.0   
2           [{'iso_639_1': 'en', 'name': 'English'}]         0.0          0.0   
3           [{'iso_639_1': 'en', 'name': 'English'}]  16000000.0   81452156.0   
4           [{'iso_639_1': 'en', 'name': 'English'}]         0.0   76578911.0   

   Runtime Release_Date                               Production_Companies  \
0     81.0   1995-10-30     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1    104.0   1995-12-15  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2    101.0   1995-12-22  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3    127.0   1995-12-22  [{'name': 'Twentieth Century Fox Film Corporat...   
4    106.0   1995-02-10  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                Production_Countries    imdb_id  Popularity  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114709   21.946943   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113497   17.015539   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113228   11.712900   
3  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0114885    3.859495   
4  [{'iso_3166_1': 'US', 'name': 'United States o...  tt0113041    8.387519   

   Vote_Average  Vote_Count  \
0           7.7      5415.0   
1           6.9      2413.0   
2           6.5        92.0   
3           6.1        34.0   
4           5.7       173.0   

                               Belongs_To_Collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...   
3                                                NaN   
4  {'id': 96871, 'name': 'Father of the Bride Col...   

                                             Tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                                            Overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                              Genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                         Genres_Parse         Belongs_To_Collection_Parse  \
0   ['Animation', 'Comedy', 'Family']            ['Toy Story Collection']   
1  ['Adventure', 'Fantasy', 'Family']                                 NaN   
2               ['Romance', 'Comedy']       ['Grumpy Old Men Collection']   
3      ['Comedy', 'Drama', 'Romance']                                 NaN   
4                          ['Comedy']  ['Father of the Bride Collection']   

    Spoken_Languages_Parse                         Production_Companies_Parse  \
0              ['English']                        ['Pixar Animation Studios']   
1  ['English', 'Français']  ['TriStar Pictures', 'Teitler Fi

Prepare the DF for Model

In [242]:
df2 = df[['Vote_Average','Vote_Count','Budget','Revenue','Runtime','Score']].copy()

#drop records if no values in these 3 columns: daszDXf[['Vote_Average','Score','Runtime']]
df2.dropna(axis="index", how="all", subset=['Vote_Average','Vote_Count','Budget','Revenue','Runtime','Score'], inplace=True)

print(df2.shape)
print(df2.head())

(44884, 6)
   Vote_Average  Vote_Count      Budget      Revenue  Runtime     Score
0           7.7      5415.0  30000000.0  373554033.0     81.0  7.639392
1           6.9      2413.0  65000000.0  262797249.0    104.0  6.819382
2           6.5        92.0         0.0          0.0    101.0  5.941282
3           6.1        34.0  16000000.0   81452156.0    127.0  5.707750
4           5.7       173.0         0.0   76578911.0    106.0  5.664068


Plot the Training and Test set
- plot a scatter matrix
- 75-25 for default train_test_split

In [249]:
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer

X = df[['Vote_Average','Vote_Count','Budget','Revenue','Runtime']]
y = df['Score']
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

Create KNN Regression: 
- option 1: Imputate X with NaN + remove NaN in rows for Y, 

In [251]:
# Step 0 - drop rows where y is NaN
not_nan_indices = ~np.isnan(y)
X = X.loc[not_nan_indices]
y = y[not_nan_indices]

# Step 1: Imputation - Replace NaN with the mean of each column
imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X)

# Step 2: Verify that there are no NaN values left in the data
print("Are there any NaN values after imputation?")
print(pd.DataFrame(X_imputed).isna().sum())  # Should output 0 for each column

# Step 3: Train/Test split
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)

# Step 4: Fit the KNN Regressor model
knnreg = KNeighborsRegressor(n_neighbors=5).fit(X_train, y_train)

# Step 5: Make predictions and evaluate the model
print("Predictions:", knnreg.predict(X_test))
print('R-squared test score: {:.3f}'.format(knnreg.score(X_test, y_test)))

Are there any NaN values after imputation?
0    0
1    0
2    0
3    0
4    0
dtype: int64
Predictions: [5.72256889 5.67225629 5.62593129 ... 5.78875609 5.63735532 5.53499975]
R-squared test score: 0.340


Option 2: Imputate both X and y with NaN.

In [252]:
# Step 1: Impute missing values in X
imputer_X = SimpleImputer(strategy='mean')
X_imputed = imputer_X.fit_transform(X)

# Step 2: Impute missing values in y
imputer_y = SimpleImputer(strategy='mean')
y_imputed = imputer_y.fit_transform(y.reshape(-1, 1)).ravel()  # ravel() to return to original shape

# Step 3: Check that there are no NaN values left in y
print("Are there any NaN values in y after imputation?")
print(np.isnan(y_imputed).sum())  # Should output 0

# Step 4: Train/Test split
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y_imputed, test_size=0.2, random_state=42)

# Step 5: Fit the KNN Regressor model
knnreg = KNeighborsRegressor(n_neighbors=5).fit(X_train, y_train)

# Step 6: Make predictions and evaluate the model
print("Predictions:", knnreg.predict(X_test))
print('R-squared test score: {:.3f}'.format(knnreg.score(X_test, y_test)))

AttributeError: 'Series' object has no attribute 'reshape'